Initial Guard Env, duplicate tensor guards

voznesenskym opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #90370
* #90130
* #90129

Adds the guard env defined here: https://docs.google.com/document/d/1VbiXkIhKy744Q7Lx2e8UUBvP_5RH_WU0FEc3VadVX4U/edit#heading=h.bk4qqlofhsq

cc @mlazos @soumith @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

voznesenskym committed:
[WIP] Initial Guard Env

wip

wip

Initial guard env

Working

[ghstack-poisoned]

voznesenskym committed:
Update on "[WIP] [Do not review] Initial Guard Env"


wip

wip

Initial guard env

Working

cc mlazos soumith yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "Initial Guard Env"


wip

wip

Initial guard env

Working

cc mlazos soumith yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym commented on test/dynamo/test_aot_autograd.py:
```diff
@@ -367,12 +367,8 @@ def guard_fail_fn(failure):
-            """At compilation time, graph 1 was compiled under the assumption that input 1 would be a duplicate of input 0, but at runtime this was not the case.  This indicates a guard bug in AOTAutograd or Dynamo, please file a bug to PyTorch.""",
-        )
+        fxx(x, y)
+        self.assertEqual(cc.frame_count, 2)
```
self nit: add the guard check

voznesenskym committed:
Update on "Initial Guard Env, duplicate tensor guards"


Adds the guard env defined here: https://docs.google.com/document/d/1VbiXkIhKy744Q7Lx2e8UUBvP_5RH_WU0FEc3VadVX4U/edit#heading=h.bk4qqlofhsq

cc mlazos soumith yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

ezyang commented on torch/fx/experimental/guard_env.py:
```diff
@@ -0,0 +1,85 @@
+# story, and this is a nice way to get access to it everywhere.
+# The only icurred mental overhead is rememebring to clear it, and that should happen right after we
+# extract guards out.
+GUARD_ENV = GuardEnv()
```
No to global. You can dynamically scope it as part of the make_fx process but manual clearing is too error prone. For precedent check out how we do decomposition table, for example.

ezyang commented on torch/_dynamo/guards.py:
```diff
@@ -33,12 +34,17 @@
 
 
+def tensor_identity_match(a, b):
+    return a is b
```
It seems like the rest of these helper functions live in utils, this one should too.

ezyang replied:
Also, it is not clear to me that you want to do this via closure vars. You can just generate the "is" directly in the Python code to be compiled into the guard, and that will be speedier (one opcode, versus a function call opcode)

voznesenskym replied:
agreed

voznesenskym committed:
Update on "Initial Guard Env, duplicate tensor guards"


Adds the guard env defined here: https://docs.google.com/document/d/1VbiXkIhKy744Q7Lx2e8UUBvP_5RH_WU0FEc3VadVX4U/edit#heading=h.bk4qqlofhsq

cc mlazos soumith yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "Initial Guard Env, duplicate tensor guards"


Adds the guard env defined here: https://docs.google.com/document/d/1VbiXkIhKy744Q7Lx2e8UUBvP_5RH_WU0FEc3VadVX4U/edit#heading=h.bk4qqlofhsq

cc mlazos soumith yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym commented on torch/_dynamo/convert_frame.py:
```diff
@@ -367,110 +366,113 @@ def _compile(
 ) -> Optional[GuardedCode]:
     output: Optional[OutputGraph] = None
 
-    # from .utils import print_once;  print_once(code.co_filename)
```
to reviewer note: There are no meaningful changes here, it is just tabbed over.

voznesenskym commented on torch/fx/experimental/guard_env.py:
```diff
@@ -0,0 +1,89 @@
+Note: All subtypes must be handled exhaustively in
+torch._dynamo.guards._parse_guard_env_guards to avoid a RuntimeError.
+"""
+class GuardEnvExpr():
```
sg

voznesenskym committed:
Update on "Initial Guard Env, duplicate tensor guards"


Adds the guard env defined here: https://docs.google.com/document/d/1VbiXkIhKy744Q7Lx2e8UUBvP_5RH_WU0FEc3VadVX4U/edit#heading=h.bk4qqlofhsq

cc mlazos soumith yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "Initial Guard Env, duplicate tensor guards"


Adds the guard env defined here: https://docs.google.com/document/d/1VbiXkIhKy744Q7Lx2e8UUBvP_5RH_WU0FEc3VadVX4U/edit#heading=h.bk4qqlofhsq

cc mlazos soumith yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

ezyang commented on torch/_dynamo/guards.py:
```diff
@@ -673,6 +675,25 @@ def combine_scopes(left, right):
+        guard_str = ""
+        for guard in guards:
+            if isinstance(guard, DuplicateInputs):
+                guard_str += f"{guard.arg_a} is {guard.arg_b}"
```
uhhh, surely you need an `and` or something here? lol

voznesenskym committed:
Update on "Initial Guard Env, duplicate tensor guards"


Adds the guard env defined here: https://docs.google.com/document/d/1VbiXkIhKy744Q7Lx2e8UUBvP_5RH_WU0FEc3VadVX4U/edit#heading=h.bk4qqlofhsq

cc mlazos soumith yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "Initial Guard Env, duplicate tensor guards"


Adds the guard env defined here: https://docs.google.com/document/d/1VbiXkIhKy744Q7Lx2e8UUBvP_5RH_WU0FEc3VadVX4U/edit#heading=h.bk4qqlofhsq

cc mlazos soumith yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "Initial Guard Env, duplicate tensor guards"


Adds the guard env defined here: https://docs.google.com/document/d/1VbiXkIhKy744Q7Lx2e8UUBvP_5RH_WU0FEc3VadVX4U/edit#heading=h.bk4qqlofhsq

cc mlazos soumith yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "Initial Guard Env, duplicate tensor guards"


Adds the guard env defined here: https://docs.google.com/document/d/1VbiXkIhKy744Q7Lx2e8UUBvP_5RH_WU0FEc3VadVX4U/edit#heading=h.bk4qqlofhsq

cc mlazos soumith yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

jansel commented on torch/_dynamo/convert_frame.py:
```diff
@@ -367,110 +366,113 @@ def _compile(
-        instructions[:] = output.output_instructions
-        code_options.update(output.code_options)
+    guard_env = GuardEnv()
+    with guarding(guard_env):
```
This with can go around the function call instead of the `def ...` part.  The env doesn't matter for a function def.  This will make this file's diff easier to read.

voznesenskym replied:
Done! I added a comment about why it is where it is too.

ezyang commented on torch/fx/experimental/guard_env.py:
```diff
@@ -0,0 +1,102 @@
+@dataclasses.dataclass
+class DuplicateInputs(GuardEnvExpr):
+    arg_a: str
+    arg_b: str
```
I think, directionally, my biggest discomfort with introducing a new GuardEnvExpr concept here that is distinct from Dynamo guards, is that Dynamo may very well WANT to generate guards like this (e.g., it is processing a conditional `a is b`) and it seems weird to have two possible ways to generate a guard (regular Guard, or GuardEnvExpr)

jansel replied:
Agreed, it would be good to consolidate how we manage guards into a single place.  That doesn't mean all guards need to be sympy guards (or dynamo guards)

voznesenskym replied:
This is due to a layering issue - aot_autograd is below dynamo, so this needs to be in another place. We could violate layering and reuse the definitions in guards.py, or we could lower those down to a shared place, or we could do something like this. This felt the more reasonable, especially given the fact that we have another guard system in shape_env already that follows a similar pattern.

voznesenskym replied:
This is also different from dynamo guards as all dynamo guards handle a single object at a time, the entire guards static_fn pattern is predicated on passing a Guard object which represent a single arg (see the `self.get` calls in guards.py). We could break that assumption and make multi-target guards, and rewrite that file, or we could carve out a special one-off guard that takes multiple tensors.

voznesenskym replied:
Fwiw, I understand the concern, and we could do something like:

1) throw out DuplicateInputs
2) Make the mapping tensor:source instead of tensor:name
3) define a new Guard type that is multi-source / or a Source type that is multi-guard (we need to model the fact that this is a 2 object guard somewhere)
4) Call Guards.DUPLICATE_INPUTS with either the 2 object guard, or with 2 guards for the 2 sources

Would you prefer that? Does that feel cleaner to you both? it does not to me.

jansel commented on torch/fx/experimental/guard_env.py:
```diff
@@ -0,0 +1,102 @@
+        self._guards.clear()
+        self._tensor_to_names.clear()
+
+CURRENT_GUARD_ENV = None
```
How are we passing the shape env to each part of the stack today?  Is it a different global variable somewhere?  I'd suggest some sort of tracing context class (and if needed just one global pointer to that) that holds pointers to these sorts of things and also allows us to guarantee they are set.

ezyang replied:
ShapeEnv is accessible from all SymInts, so all the points when we may want to register guards (e.g. an equality test on symint) can just directly add the guard to the SymInt. So ShapeEnv didn't really need a context class (although maybe it should.)

voznesenskym replied:
We were talking about global vs passed around and Ed suggested this dynamic scoping to be a medium between the two. I don't have strong opinions on either way, but the plumbing story right now to backends is not amazing (as you recall from the kwargs/pass fake_mode around PRs we tried and failed prior). 

There's a few ways we could take this:

1) Leave as is, its a one off for now, until another thing with a similar lifecycle and pass-through-backends story needs to be developed
2) Match the fake_mode pattern more closely, and ride guard_env down on the inputs (I dislike it, but its worth mentioning for consistency)
3) Make a tracing context we pass around, move the shape_env and guard_env there

voznesenskym committed:
Update on "Initial Guard Env, duplicate tensor guards"


Adds the guard env defined here: https://docs.google.com/document/d/1VbiXkIhKy744Q7Lx2e8UUBvP_5RH_WU0FEc3VadVX4U/edit#heading=h.bk4qqlofhsq

cc mlazos soumith yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]
